pkg/storage/allocator_scorer.go

```diff
@@ -697,63 +705,74 @@ func constraintCheck(store roachpb.StoreDescriptor, constraints config.Constrain
 	return true, positive
 }
 
-// diversityScore returns a score between 1 and 0 where higher scores are stores
-// with the fewest locality tiers in common with already existing replicas.
-func diversityScore(
+// diversityAllocateScore returns a score between 1 and 0. It is only used for allocateCandidates.
+// A much higher score just means the store may be a better fit for allocate taget.
+func diversityAllocateScore(
 	store roachpb.StoreDescriptor, existingNodeLocalities map[roachpb.NodeID]roachpb.Locality,
 ) float64 {
-	minScore := roachpb.MaxDiversityScore
+	var sumScore float64
+	var numSamples int
+	// We didn't calculate the overall diversityScore for the range, because the original overall diversityScore
+	// of this range is always the same.
 	for _, locality := range existingNodeLocalities {
-		if newScore := store.Node.Locality.DiversityScore(locality); newScore < minScore {
-			minScore = newScore
-		}
+		newScore := store.Node.Locality.DiversityScore(locality)
+		sumScore += newScore
+		numSamples++
+	}
+	// When this range has no replica, it means any node will be a perfect fit for this range.
+	if numSamples == 0 {
+		return 1.0
 	}
-	return minScore
+	return sumScore / float64(numSamples)
 }
 
-// diversityRemovalScore is the same as diversityScore, but for a node that's
-// already present in existingNodeLocalities. It works by calculating the
-// diversityScore for nodeID as if nodeID didn't already have a replica.
-// As with diversityScore, a higher score indicates that the node is a better
-// fit for the range (i.e. keeping it around is good for diversity).
+// diversityRemovalScore calculates the average diversityScore if we try to remove the replica from this node.
+// A higher score indicates that the node is a better fit for the range (i.e. keeping it around is good for diversity).
 func diversityRemovalScore(
 	nodeID roachpb.NodeID, existingNodeLocalities map[roachpb.NodeID]roachpb.Locality,
 ) float64 {
-	minScore := roachpb.MaxDiversityScore
+	var sumScore float64
+	var numSamples int
 	locality := existingNodeLocalities[nodeID]
+	// We didn't calculate the overall diversityScore for the range, because the original overall diversityScor
+	// of this range is always the same.
 	for otherNodeID, otherLocality := range existingNodeLocalities {
 		if otherNodeID == nodeID {
 			continue
 		}
-		if newScore := otherLocality.DiversityScore(locality); newScore < minScore {
-			minScore = newScore
-		}
+		newScore := locality.DiversityScore(otherLocality)
```

Comment:
Hmm... How about I add the comment like

```
We have no need the calculate the overall diversityScore after we remove the replica on `nodeID`, because the overall diversityScore of this range is always the same, we just calculate the diversityScore between the `nodeID` and other replicas for this range.
```

Reply:
I don't know how you choose to think about it, but my intuition around why taking this shortcut is ok is that it's because the replica with the worst average diversityScore is the one that's dragging down the range's overall diversityScore the most.

Reply:
Yeah, something like that.